### PR TITLE
llama-cpp: fix build due to openblas update; 1483 -> 1538

### PR DESCRIPTION
**Pull request by ghost**
pkg-confg file for openblas moved from openblas.pc -> openblas64.pc
    
add patch so build checks for openblas64 module. submitted patch upstream  https://github.com/ggerganov/llama.cpp/pull/4134
    
 https://github.com/NixOS/nixpkgs/pull/255443 updated from openblas  0.3.21 -> 0.3.24.  openblas 0.3.22 moved openblas.pc -> openblas64.pc

https://github.com/OpenMathLib/OpenBLAS/issues/3790

updated the package as well.

https://hydra.nixos.org/build/241259310
ZHF: #265948

Result of `nixpkgs-review pr 268449` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>llama-cpp</li>
    <li>ollama</li>
  </ul>
</details>

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by elohmeier** [comment]
LGTM

**Review by elohmeier — APPROVED** [review]

File: pkgs/by-name/ll/llama-cpp/package.nix [hunk 1]
```diff
@@ -37,15 +38,25 @@ let
...
-    hash = "sha256-TYklPkqwXLt+80FSHBDA2r3xTXlmgqB7sOt2mNnVNso=";
+    hash = "sha256-3JPGKJbO7Z3Jxz9KNSLYBAM7zQ+RJwBqsfRtpK6JS48=";
   };
 
+  patches = [
+    # openblas > v0.3.21 64-bit pkg-config file is now named openblas64.pc
+    # can remove when patch is accepted upstream
+    (fetchpatch {
+      name = "openblas64-pkg-config.patch";
+      url = "https://github.com/ggerganov/llama.cpp/pull/4134/commits/c885cc9f76c00557601b877136191b0f7aadc320.patch";
```
Comment (Mic92) [inline pkgs/by-name/ll/llama-cpp/package.nix:55]: Will this patch stay valid if the pull request is based and the commit is no longer part of the pull request?

Comment (Mic92) [reply pkgs/by-name/ll/llama-cpp/package.nix:55]: If not than:

```
# https://github.com/ggerganov/llama.cpp/pull/4134
url = "https://github.com/ggerganov/llama.cpp/commit/c885cc9f76c00557601b877136191b0f7aadc320.patch";
```

is the better choice.

Comment (a-n-n-a-l-e-e) [reply pkgs/by-name/ll/llama-cpp/package.nix:55]: thanks -- i thought i fixed that -- obviously not.

File: pkgs/by-name/ll/llama-cpp/package.nix [hunk 2]
```diff
@@ -37,15 +38,25 @@ let
...
     owner = "ggerganov";
     repo = "llama.cpp";
     rev = "refs/tags/b${finalAttrs.version}";
-    hash = "sha256-TYklPkqwXLt+80FSHBDA2r3xTXlmgqB7sOt2mNnVNso=";
+    hash = "sha256-3JPGKJbO7Z3Jxz9KNSLYBAM7zQ+RJwBqsfRtpK6JS48=";
   };
 
+  patches = [
+    # openblas > v0.3.21 64-bit pkg-config file is now named openblas64.pc
+    # can remove when patch is accepted upstream
```
Comment (Mic92) [inline pkgs/by-name/ll/llama-cpp/package.nix:52]:
```suggestion
    # can remove when patch is accepted upstream (https://github.com/ggerganov/llama.cpp/pull/4134)
```

**Comment by a-n-n-a-l-e-e** [comment]
thanks!

**Comment by a-n-n-a-l-e-e** [comment]
@elohmeier I had mistakenly assumed that the 64 was revering to pointer size when i made this change -- it doesn't. the 64 refers to the integer size, so `openblas64.pc` is paired with the ILP64 openblas, and `openblas.pc` is paired with the LP64.

is this package trying to use the ILP64 openblas or the LP64 openblas?

the [nix flake](https://github.com/ggerganov/llama.cpp/blob/07178c98e1b61a5e2af39d347add12e7eb9e08e1/flake.nix) also uses the ILP64 openblas -- there was a [comment]( https://github.com/ggerganov/llama.cpp/pull/4134#issuecomment-1817908396)
in my upstream PR which made me think that the LP64 was expected but I think i misunderstood.

**Comment by elohmeier** [comment]
TBH I'm not using the OpenBLAS implementation personally and did not test it in depth. I don't really have an opinion on which one to use here, but would suggest to copy the upstream choice from their nix flake.

**Pull request merged**